cmd/helmwave/main.go

```diff
@@ -120,3 +128,54 @@ func version() *cli.Command {
 		},
 	}
 }
+
+func cancelFlag() *cli.BoolFlag {
+	return &cli.BoolFlag{
+		Name:  "cancel-on-kill",
+		Usage: "cancel helm on SigINT,SigTERM",
+
+		Value:   false,
+		EnvVars: []string{"HELMWAVE_HANDLE_SIGNAL"},
+		Hidden:  true, // experimental
+	}
+}
+
+type signalCtx struct {
+	context.Context
+}
+
+func (ctx signalCtx) Err() error {
+	// Unwrap for readablity, if ctx cancelled by reciving a signal
+	if sigErr, ok := context.Cause(ctx.Context).(signalErr); ok {
+		return sigErr
+	}
+	return ctx.Context.Err()
+}
+
+type signalErr struct {
+	sig os.Signal
+}
+
+func (e signalErr) Error() string {
+	return "context canceled: got signal " + e.sig.String()
+}
+
+// cancelCtxOnSignal closes Done channel when one of the listed signals arrives
+func cancelCtxOnSignal(parent context.Context, signals ...os.Signal) (ctx context.Context) {
+	ctx, cancel := context.WithCancelCause(parent)
+	ctx = signalCtx{ctx} // wrap for err handling
```

Comment:
I think in case of signal we don't really require every `context canceled` line to be `context canceled: got signal blabla`.

So instead of wrapping context into another structure and passing it through the whole program it would be better to use `context.AfterFunc` and print the real cause of context cancelling just once.